### PR TITLE
RevoEx SO, GameSpy library

### PR DESCRIPTION
**Pull request by wait-wtf**
```
[DOL] Verifying...
[DOL] Everything went okay! Output is matching! ^^
[REL] Everything went okay! Output is matching! ^^
[DOL] 9.254% code, 1.704% data decompiled
  -> [RVL] 1.924% code, % data decompiled
  -> [EGG] 4.713% code, % data decompiled
  -> [SPY] 84.804% code, % data decompiled
[REL] 0.012% code, 0.005% data decompiled
--- main.dol + StaticR.rel ---
 3.622% code, 1.352% data decompiled
------
Player:
 - 5462 BR (main.dol)
 - 5000 VR (StaticR.rel)
 ```

Requires #22, #23

### CODE REVIEW

**Comment by wait-wtf** [comment]
@riidefi This was about as far as I could push it. It's ready for review now.

File: source/gamespy/common/revolution/gsThreadRevolution.c [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+#include <stdbool.h>
+
+// Begin of Threading for Revolution
+///////////////////////////////////////////////////////////////////////////////
+///////////////////////////////////////////////////////////////////////////////
+gsi_u32 gsiInterlockedIncrement(gsi_u32* value) {
+  int enabled = OSDisableInterrupts();
+
+  gsi_u32 ret = ++(*value);
+  OSRestoreInterrupts(enabled);
```
Comment (riidefi) [inline source/gamespy/common/revolution/gsThreadRevolution.c:14]: Just a note: there's a better way to do this without disabling interrupts. If we ever wanted to enable multiple WiiU cores, these OSDisableInterrupts calls would get in our way.

File: source/gamespy/ghttp/ghttpBuffer.c [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+  while (newLen >= buffer->size) {
+    // Check for a fixed buffer.
+    ////////////////////////////
+    if (buffer->fixed) {
+      buffer->connection->completed = GHTTPTrue;
+      buffer->connection->result = GHTTPBufferOverflow;
+      return GHTTPFalse;
+    }
+
+    bResult = ghiResizeBuffer(buffer, buffer->sizeIncrement);
```
Comment (riidefi) [inline source/gamespy/ghttp/ghttpBuffer.c:233]: But what if dataLen > buffer->sizeIncrement..?

File: source/gamespy/ghttp/ghttpCommon.c [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+      connection->decodeBuffer.len = bytesToKeep;
+    }
+  }
+
+  return GHTTPTrue;
+}
+
+// Receive some data.
+/////////////////////
+asm GHIRecvResult ghiDoReceive(GHIConnection* connection, char buffer[],
```
Comment (riidefi) [inline source/gamespy/ghttp/ghttpCommon.c:122]: Can you add a NON_MATCHING version of these?

File: source/gamespy/qr2/qr2.c [hunk 4]
```diff
@@ -0,0 +1,1605 @@
...
+  foo();
+  x = "CanReceiveOnSocket() returned true, but recvfrom failed!\r\n"; // 0x8027d108
+  foo = (void (*)(void))x;
+  foo();
+}
+
+char qr2_check_queries_indata[INBUF_LEN];
+
+// clang-format off
+asm void qr2_check_queries(qr2_t qrec) {
```
Comment (riidefi) [inline source/gamespy/qr2/qr2.c:459]: Add NON_MATCHING version if possible

**Review by riidefi — CHANGES_REQUESTED** [review]
Incredible work. Just a few notes. :)

File: source/gamespy/ghttp/ghttpBuffer.c (hunk 2 above)
Comment (SeekyCt) [reply source/gamespy/ghttp/ghttpBuffer.c:233]: Won't it keep increasing by sizeIncrement until it fits or runs out of memory?

Comment (riidefi) [reply source/gamespy/ghttp/ghttpBuffer.c:233]: Yep

File: source/gamespy/GP/gp.c [hunk 5]
```diff
@@ -0,0 +1,3107 @@
...
+  //////////////////////////
+  if (iconnection->connectState == GPI_DISCONNECTED)
+    Error(connection, GP_PARAMETER_ERROR,
+          "The connection has already been disconnected.");
+
+  pProfile = gpiFindBuddy(connection, index);
+  if (!pProfile)
+    Error(connection, GP_PARAMETER_ERROR, "Invalid index.");
+
+  if (pProfile->buddyStatus)
```
[the remark this answers is not in the document]
Comment (wait-wtf) [reply source/gamespy/GP/gp.c:1748]: Fixed

File: source/gamespy/ghttp/ghttpCommon.c (hunk 3 above)
Comment (wait-wtf) [reply source/gamespy/ghttp/ghttpCommon.c:221]: Added NON_MATCHING version of ghiDoReceive().

File: source/gamespy/qr2/qr2.c (hunk 4 above)
Comment (wait-wtf) [reply source/gamespy/qr2/qr2.c:478]: Added NON_MATCHING version of qr2_check_queries().

**Comment by wait-wtf** [comment]
See also: https://github.com/riidefi/mkw/issues/28

**Pull request merged**